File: netlogo-gui/src/main/nvm/Argument.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm;

import org.nlogo.agent.AgentSet;
import org.nlogo.api.Dump;
import org.nlogo.api.TypeNames;
import org.nlogo.core.Syntax;
import java.util.List;

/**
 * Passes arguments to extension primitives.
 */
public strictfp class Argument
    implements org.nlogo.api.Argument {
  private final Context context;
  protected final Reporter arg;
  private Object cached = null;

  public Argument(Context context, Reporter arg) {
    this.context = context;
    this.arg = arg;
  }

  public Object get()
      throws org.nlogo.api.LogoException {
    if (cached == null) {
      cached = arg.report(context);
    }
    if (cached instanceof org.nlogo.agent.Agent &&
        ((org.nlogo.agent.Agent) cached).id == -1) {
      cached = org.nlogo.core.Nobody$.MODULE$;
    }
    return cached;
  }

  public AgentSet getAgentSet()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (AgentSet) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException
          (getExceptionMessage
           (Syntax.AgentsetType(), obj));
    }
  }


  public org.nlogo.api.Agent getAgent()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Agent) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.AgentType(), obj));
    }
  }

  public Boolean getBoolean()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (Boolean) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.BooleanType(), obj));
    }
  }

  public boolean getBooleanValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Boolean) obj).booleanValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.BooleanType(), obj));
    }
  }

  public double getDoubleValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Double) obj).doubleValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.NumberType(), obj));
    }
  }

  public int getIntValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Double) obj).intValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.NumberType(), obj));
    }
  }

  public org.nlogo.core.LogoList getList()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.core.LogoList) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.ListType(), obj));
    }
  }

  public org.nlogo.api.Patch getPatch()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Patch) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.PatchType(), obj));
    }
  }

  public String getString()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (String) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.StringType(), obj));
    }
  }


  public org.nlogo.api.Turtle getTurtle()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Turtle) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.TurtleType(), obj));
    }
  }

  public org.nlogo.api.Link getLink()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Link) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.LinkType(), obj));
    }
  }

  public org.nlogo.api.AnonymousReporter getReporter()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.AnonymousReporter) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.ReporterType(), obj));
    }
  }

  public org.nlogo.api.AnonymousCommand getCommand()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.AnonymousCommand) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.CommandType(), obj));
    }
  }

  @SuppressWarnings("unchecked") public List<org.nlogo.core.Token> getCode()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (List<org.nlogo.core.Token>) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.CodeBlockType(), obj));
    }
  }

  public org.nlogo.core.Token getSymbol()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.core.Token) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.SymbolType(), obj));
    }
  }

  // if you're looking for the cities extension's <code>getReporter</code>
  // method, that has been removed. Change the extension and use
  // <code>getSymbol</code> instead.

  private String getExceptionMessage(int wantedType, Object badValue) {
    String result = "Expected this input to be "
        + TypeNames.aName(wantedType) + " but got "
        + (badValue == org.nlogo.core.Nobody$.MODULE$
        ? "NOBODY"
        : "the " + TypeNames.name(badValue)
        + " " + Dump.logoObject(badValue))
        + " instead.";

    return result;

  }

}

File: netlogo-core/src/main/nvm/Argument.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm

import org.nlogo.{ api, core },
  core.{ Nobody, Reference, Syntax, Token, TypeNames },
  api.{ Dump, ExtensionException, LogoException }

import java.util.{ List => JList }

import scala.unchecked

/**
 * Passes arguments to extension primitives.
 */
class Argument(context: Context, arg: Reporter) extends api.Argument {

  private var cached: AnyRef = null

  def get = {
    unsafeGet match {
      case a: api.Agent if a.id == -1 =>
        cached = Nobody
      case _ =>
    }
    cached
  }

  // This is not safe when the argument might be an agent but is much faster
  // when we know we don't want an agent. -- BCH 06/28/2017
  private def unsafeGet = {
    if (cached == null)
      cached = arg.report(context)
    cached
  }

  @throws(classOf[ExtensionException])
  def getAgentSet: api.AgentSet =
    unsafeGet match {
      case agents: api.AgentSet => agents
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.AgentsetType, x))
    }

  @throws(classOf[ExtensionException])
  def getAgent: api.Agent =
    get match {
      case agent: api.Agent => agent
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.AgentType, x))
    }

  @throws(classOf[ExtensionException])
  def getBoolean: java.lang.Boolean =
    unsafeGet match {
      case b: java.lang.Boolean => b
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.BooleanType, x))
    }

  @throws(classOf[ExtensionException])
  def getBooleanValue: Boolean =
    unsafeGet match {
      case b: java.lang.Boolean => b.booleanValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.BooleanType, x))
    }

  @throws(classOf[ExtensionException])
  def getDouble: java.lang.Double =
    unsafeGet match {
      case d: java.lang.Double => d
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.NumberType, x))
    }

  @throws(classOf[ExtensionException])
  def getDoubleValue: Double =
    unsafeGet match {
      case d: java.lang.Double => d.doubleValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.NumberType, x))
    }

  @throws(classOf[ExtensionException])
  def getIntValue: Int =
    unsafeGet match {
      case d: java.lang.Double => d.intValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.NumberType, x))
    }

  @throws(classOf[ExtensionException])
  def getList: core.LogoList =
    unsafeGet match {
      case l: core.LogoList => l
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.ListType, x))
    }

  @throws(classOf[ExtensionException])
  def getPatch: api.Patch =
    get match {
      case p: api.Patch => p
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.PatchType, x))
    }

  @throws(classOf[ExtensionException])
  def getString: String =
    unsafeGet match {
      case s: String => s
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.StringType, x))
    }

  @throws(classOf[ExtensionException])
  def getTurtle: api.Turtle =
    get match {
      case t: api.Turtle => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.TurtleType, x))
    }

  @throws(classOf[ExtensionException])
  def getLink: api.Link =
    get match {
      case l: api.Link => l
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.LinkType, x))
    }

  @throws(classOf[ExtensionException])
  def getReporter: api.AnonymousReporter =
    unsafeGet match {
      case t: api.AnonymousReporter => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.ReporterType, x))
    }

  @throws(classOf[ExtensionException])
  def getCommand: api.AnonymousCommand =
    unsafeGet match {
      case t: api.AnonymousCommand => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.CommandType, x))
    }


  @throws(classOf[ExtensionException])
  def getCode: JList[Token] =
    unsafeGet match {
      case t: JList[Token] @unchecked => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.CodeBlockType, x))
    }

  @throws(classOf[ExtensionException])
  def getSymbol: Token =
    unsafeGet match {
      case t: Token => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.SymbolType, x))
    }


  /**
   * Used by the GIS extension. Other extensions should use
   * getSymbol where possible. The compiler turns the reference into a _const,
   * we just ensure the value is a reference.
   */
  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getReference: Reference =
    unsafeGet match {
      case ref: Reference => ref
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.ReferenceType, x))
    }

  /**
   * Used by the rnd extension. Other extensions should use getReporter where possible
   */
  def unevaluatedArgument: Reporter = arg

  // if you're looking for the cities extension's <code>getReporter</code>
  // method, that has been removed. Change the extension and use
  // <code>getSymbol</code> instead.

  def getExceptionMessage(wantedType: Int, badValue: AnyRef) =
    "Expected this input to be " +
      TypeNames.aName(wantedType) + " but got " +
       (if(badValue == Nobody) "NOBODY"
        else ("the " + api.TypeNames.name(badValue) + " " +
              Dump.logoObject(badValue)) +
              " instead.")

}
